llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp

```diff
@@ -318,12 +317,11 @@ static unsigned getShufflevectorNumGroups(ArrayRef<Value *> VL) {
           int Index;
           if (!SV->isExtractSubvectorMask(Index))
             return false;
-          for (int I : seq<int>(Index, Index + SV->getShuffleMask().size()))
-            ExtractionIndex.push_back(I);
+          UsedIndex.set(Index, Index + SV->getShuffleMask().size());
           return true;
         }))
       return 0;
-    if (!is_sorted(ExtractionIndex))
+    if (!UsedIndex.all())
```

Comment:
What if we just reuse the same group several times but some other group do not use at all?

Reply:
Do you have an example? I am not sure what you mean.

Reply:
I mean, <0,1> extracted several time, while <2,3> is not extracted at all

Reply:
Cannot. We had a `getShufflevectorNumGroups` in `getScalarsVectorizationState`. If `getShufflevectorNumGroups` returns 0, then `buildTree_rec` will stop (a `TreeEntry::NeedToGather` will return).
`getShufflevectorNumGroups` use
```
    SmallVector<int> ExtractionIndex(SVNumElements);
    if (!all_of(Group, [&](Value *V) {
          auto *SV = cast<ShuffleVectorInst>(V);
          // From the same source.
          if (SV->getOperand(0) != Src)
            return false;
          int Index;
          if (!SV->isExtractSubvectorMask(Index))
            return false;
          for (int I : seq<int>(Index, Index + SV->getShuffleMask().size()))
            ExtractionIndex.push_back(I);
          return true;
        }))
      return 0;
    if (!is_sorted(ExtractionIndex))
      return 0;
```
to make sure each elements will be accessed.